pageserver/src/tenant.rs

```diff
@@ -3951,28 +4036,96 @@ impl Tenant {
+                pausable_failpoint!("timeline-creation-after-uninit");
+                Ok(StartCreatingTimelineResult::CreateGuard(create_guard))
+            }
+            Err(TimelineExclusionError::AlreadyCreating) => {
+                // Creation is in progress, we cannot create it again, and we cannot
+                // check if this request matches the existing one, so caller must try
+                // again later.
+                Err(CreateTimelineError::AlreadyCreating)
+            }
+            Err(TimelineExclusionError::Other(e)) => Err(CreateTimelineError::Other(e)),
+            Err(TimelineExclusionError::AlreadyExists { existing, arg }) => {
+                {
+                    let existing = &existing.create_idempotency;
+                    let _span = info_span!("idempotency_check", ?existing, ?arg).entered();
+
+                    match (existing, &arg) {
+                        // FailWithConflict => no idempotency check
+                        (CreateTimelineIdempotency::FailWithConflict, _)
+                        | (_, CreateTimelineIdempotency::FailWithConflict) => {
+                            warn!("timeline already exists, failing request");
+                            return Err(CreateTimelineError::Conflict);
+                        }
+                        // Idempotent <=> CreateTimelineIdempotency is identical
+                        (x, y) if x == y => {
+                            info!("timeline already exists and idempotency matches, succeeding request");
+                            // fallthrough
+                        }
+                        (_, _) => {
+                            warn!("idempotency conflict, failing request");
+                            return Err(CreateTimelineError::Conflict);
+                        }
+                    }
+                }
+
+                // Wait for uploads to complete, so that when we return Ok, the timeline
+                // is known to be durable on remote storage. Just like we do at the end of
+                // this function, after we have created the timeline ourselves.
+                //
+                // We only really care that the initial version of `index_part.json` has
+                // been uploaded. That's enough to remember that the timeline
+                // exists. However, there is no function to wait specifically for that so
+                // we just wait for all in-progress uploads to finish.
+                existing
+                    .remote_client
+                    .wait_completion()
+                    .await
+                    .context("wait for timeline uploads to complete")?;
+
+                // TODO: shouldn't we also wait for timeline to become active?
+                // Code before this(https://github.com/neondatabase/neon/pull/9366) refactoring
+                // didn't do it.
+
+                Ok(StartCreatingTimelineResult::Idempotent(existing))
```

Comment:
Hm, about this piece of code here:

1. Can we rely on the callers to do wait_completion()? The not-cfg(test) code always waits for completion already.
2. About the `TODO :shouldn't we also wait...`: this would probably also make more sense in the place that does the activation in the case of `StartCreatingTimelineResult::Created`.

Reply:
> Hm, about this piece of code here:
> 
>     1. Can we rely on the callers to do wait_completion()? The not-cfg(test) code always waits for completion already.

Yes, that would be my expectation.

>     2. About the `TODO :shouldn't we also wait...`: this would probably also make more sense in the place that does the activation in the case of `StartCreatingTimelineResult::Created`.

Yeah, I think it makes sense to move to the place where we do the activation.

Question though:
For `StartCreatingTimelineResult::Idempotent`, why do we call `wait_completion` on the remote client twice?
Once in `start_creating_timeline` and once in `create_timeline`?